cmd/otel-allocator/allocation/per_node.go

```diff
@@ -0,0 +1,295 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package allocation
+
+import (
+	"sync"
+
+	"github.com/go-logr/logr"
+	"github.com/open-telemetry/opentelemetry-operator/cmd/otel-allocator/diff"
+	"github.com/open-telemetry/opentelemetry-operator/cmd/otel-allocator/target"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/model"
+)
+
+var _ Allocator = &perNodeAllocator{}
+
+const (
+	perNodeStrategyName = "per-node"
+
+	podNodeNameLabel model.LabelName = "__meta_kubernetes_pod_node_name"
```

Comment:
One doubt I'm having is how to make the label matching more robust, I'm mainly thinking about:
- Which service discoveries should we be targeting here? For service / pod monitors, we should be able depend on `__meta_kubernetes_pod_node_name`, but I wonder if there might be cases where this label is not available (perhaps due to using different role for Kubernetes SD or using different SD? Has anyone encountered people using other SDs?)
- How to handle targets that we might not match against any node (such as static configs), currently I'm skipping those. I'm thinking if that's fine (maybe should be documented?) or we want to have some type of fallback to distribute such targets, if that makes sense?

Reply:
For the labels, we are constrained by the fact that the value needs to match the collector's Node field. So we can't do much better than Kubernetes Service Discovery, or at least not without significant additional complexity. So what I'd do there is just check all the Node fields from SD:

* `__meta_kubernetes_node_name`
* `__meta_kubernetes_pod_node_name`
* `__meta_kubernetes_endpoint_node_name`

The problem of assigning targets without a Node doesn't have an obvious solution. I can see the following options:

* Ignore these targets
* Assign them randomly
* Assign them according to some deterministic scheme (could be what the consistent hashing strategy uses)

We should discuss this in the issue, rather than here imo.